### ISSUE TITLE
Compare with bert-as-service

### ISSUE THREAD

**Issue by linkinbird** [zhihu/cuBERT#25]
We are currently using [bert-as-service](https://github.com/hanxiao/bert-as-service) done by Tencent AI lab which also support inference on multiple GPUs (output feature embedding layer) .

Will cuBERT be faster? Can you add benchmarks for that?
Thanks

**Comment by levyfan** [comment]
They use tensorflow. We are much faster than tensorflow. You can have a try.

**Comment by linkinbird** [comment]
@levyfan I tested the python api
in cuBERT_test.py the first test:
`model.compute(input_ids, input_mask, segment_ids, output, output_type=output_type)`

if output_type =0 (LOGITS), output is [-2.9427543, -1.4876307]
if output_type = 1 (POOLED_OUTPUT), output is [-0.54790974, -0.02262498]
if output_type = 2 (SEQUENCE_OUTPUT), output is array([], shape=(-4608227147052022383,), dtype=float32)
if output_type = 3 (EMBEDDING_OUTPUT), output is [0.15934479, 0.21887018]

I'm not sure if i was using it right. Because for each call i have to restart the python or i'll get different kinds of errors. So it's impossible to do compute in a 'while' loop for online serving.

And sometimes i'll get Segmentation fault: 11, or Segmentation fault: 7 even i already restarted the python. 
Sorry i can't debug in C++ code, can you provide more detailed examples for python api? 

Thanks

**Comment by levyfan** [comment]
You need to provide enough space for different output_type as in https://github.com/zhihu/cuBERT/blob/93143d2f7c7b40eb10b1aed542628ef87e7031fd/python/cuBERT_test.py#L26

* for output_type =0 (LOGITS), it should be [batch_size, num_labels]. In our provided test model, num_labels=1.

* for output_type = 1 (POOLED_OUTPUT), it should be [batch_size, hidden_size].

* for output_type = 2 (SEQUENCE_OUTPUT), it should be [batch_size, seq_length, hidden_size].

**Comment by levyfan** [comment]
And also we have fixed several bugs, please checkout our latest release https://github.com/zhihu/cuBERT/releases/tag/v0.0.2

For easy python benchmark, please have a look at https://github.com/zhihu/cuBERT/blob/master/python/cuBERT_benchmark.py

**Comment by linkinbird** [comment]
@levyfan Test done, X2 fast !!

**Comment by linkinbird** [comment]
@levyfan the python api tokenize_compute() has to take sequence pairs as input. How can i use it as single sequences mode?

In src/cuBERT.cpp
void cuBERT_tokenize_compute will check if text_b == nullptr

But in python i can't set text_b as nullptr.
I'll get errors if i skip text_b or set text_b to [] or "" in python

**Comment by levyfan** [comment]
Just set text_b to `None`.

```
model.tokenize_compute(text_a, None, output, output_type=output_type)
```

**Comment by linkinbird** [comment]
I'm testing it for FAQ matching.
If i use the POOLED_OUTPUT as features, it seems the second-to-last hidden layer is better than the last layer which is supported by [bert-as-service](https://bert-as-service.readthedocs.io/en/latest/section/faq.html#why-not-the-last-hidden-layer-why-second-to-last). It will be nice to have that setting.

I have not tested the fine tuned model, guess it will be OK to use the last layer.

**Comment by levyfan** [comment]
You can simply pass `num_hidden_layers` as what you want. Second-to-last should be 11 for standard BERT, and 23 for large BERT.

**Comment by zhyon404** [comment]
@linkinbird Did you try it with -fp16 and -xla? According to bert-as-service's document, it will be 1.5 times faster than without it.
